source/extensions/filters/network/kafka/serialization.h

```diff
@@ -0,0 +1,776 @@
+#pragma once
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
+
+#include "common/common/byte_order.h"
+#include "common/common/fmt.h"
+
+#include "extensions/filters/network/kafka/kafka_types.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+
+/**
+ * Deserializer is a stateful entity that constructs a result of type T from bytes provided
+ * It can be feed()-ed data until it is ready, filling the internal store
+ * When ready(), it is safe to call get() to transform the internally stored bytes into result
+ * Further feed()-ing should have no effect on a buffer (should return 0 and not move
+ * buffer/remaining)
+ * @param T type of deserialized data
+ */
+template <typename T> class Deserializer {
```

Comment:
I still do wonder if this code has overlaps with what we are doing with Thrift/MySQL/etc. I feel like we end up rewriting variants of this each time. I don't know if that's exactly true here or not, but it would be nice to one day figure that out. This is probably non-actionable but it might be worth a quick look and a TODO. cc @zuercher @rshriram